### PR TITLE
Migrate more Material framework tests to null safety.

### PR DESCRIPTION
**Pull request by darrenaustin**
## Description

Still more migration of Material framework tests to null safety.

### CODE REVIEW

File: packages/flutter/lib/src/material/switch.dart [hunk 1]
```diff
@@ -350,7 +356,7 @@ class _SwitchState extends State<Switch> with TickerProviderStateMixin {
             activeTrackColor: activeTrackColor,
             inactiveTrackColor: inactiveTrackColor,
             configuration: createLocalImageConfiguration(context),
-            onChanged: widget.onChanged,
+            onChanged: enabled ? _handleValueChanged : null,
```
Comment (goderbauer) [inline packages/flutter/lib/src/material/switch.dart:359]: _RenderSwitch.onChanged and _SwitchRenderObjectWidget.onChanged should probably also be changed to `ValueChanged<bool>?` to indicate that the value here can never be null. After that, we should just pass widget.onChanged down unchanged and add the logic from _handleValueChanged to one of the lower levels.

File: packages/flutter/lib/src/material/slider_theme.dart [hunk 2]
```diff
@@ -870,14 +870,14 @@ abstract class SliderComponentShape {
     Offset center, {
     required Animation<double> activationAnimation,
     required Animation<double> enableAnimation,
-    bool isDiscrete,
+    bool? isDiscrete,
```
Comment (goderbauer) [inline packages/flutter/lib/src/material/slider_theme.dart:873]: What does it mean if `isDiscrete` is null? We should document that.

File: packages/flutter/lib/src/material/slider_theme.dart [hunk 3]
```diff
@@ -870,14 +870,14 @@ abstract class SliderComponentShape {
...
     required TextPainter labelPainter,
     required RenderBox parentBox,
     required SliderThemeData sliderTheme,
     required TextDirection textDirection,
     required double value,
-    double textScaleFactor,
-    Size sizeWithOverflow,
+    double? textScaleFactor,
+    Size? sizeWithOverflow,
```
Comment (goderbauer) [inline packages/flutter/lib/src/material/slider_theme.dart:880]: Same with these. There's also something strange here. Some sub-classes definitely expect these to be non-null...

File: packages/flutter/lib/src/material/switch.dart (hunk 1 above)
Comment (darrenaustin) [reply packages/flutter/lib/src/material/switch.dart:359]: Yeah, I wasn't sure how far to take this change, so I stuck with just the public facing API, as that was the important one. I will look at migrating this change down to the lower levels.

File: packages/flutter/lib/src/material/slider_theme.dart (hunk 3 above)
Comment (darrenaustin) [reply packages/flutter/lib/src/material/slider_theme.dart:880]: Yeah, I wasn't sure what to do with these. If they are required and non-null then defaults need to be specified for them. It wasn't obvious to me what those defaults should be. I'll talk with Anthony and see if he can give me some guidance here.

Comment (clocksmith) [reply packages/flutter/lib/src/material/slider_theme.dart:880]: I cant see any reason for these 3 params not to be `required`

Comment (darrenaustin) [reply packages/flutter/lib/src/material/slider_theme.dart:880]: K. Thanks for the reply. I will go ahead and update those to `required`.

**Comment by pennzht** [comment]
Thank you for posting this PR. On the `ValueChanged<bool?>` issue:
It looks like there is more that should be changed. Below is a list of widgets that are affected.

* class Switch extends StatefulWidget
* class _SwitchRenderObjectWidget extends LeafRenderObjectWidget
* class _RenderSwitch extends RenderToggleable
* class SwitchListTile extends StatelessWidget
* abstract class RenderToggleable extends RenderConstrainedBox

`_RenderSwitch`, `_SwitchRenderObjectWidget`, and `RenderToggleable` are affected, because they all have an `onChanged` member variable that is passed in from `Switch`. As a result, their types must be changed to `ValueChanged<bool>?` as well.

**This raises a serious issue**: `RenderToggleable` is not only used by `Switch`; it is also used by `_RenderRadio` (radio.dart) and `_RenderCheckbox` (checkbox.dart), **both of which are possibly tristate**. However, **`Switch` is not tristate**, so there is a conflict.

In other words: `Switch`, `Radio`, and `Checkbox` all rely on `RenderToggleable`. But some are tristateable and `Switch` is not, so they disagree on what type `onChanged` should take in `RenderToggleable`.

List of changes to be made:

```
packages/flutter/lib/src/material/switch.dart L152
(class Switch extends StatefulWidget)
-  final ValueChanged<bool?>? onChanged;
+  final ValueChanged<bool>? onChanged;

packages/flutter/lib/src/material/switch.dart L443
(class _SwitchRenderObjectWidget extends LeafRenderObjectWidget)
-  final ValueChanged<bool?>? onChanged;
+  final ValueChanged<bool>? onChanged;

packages/flutter/lib/src/material/switch.dart L516
(class _RenderSwitch extends RenderToggleable)
-    required ValueChanged<bool?>? onChanged,
+    required ValueChanged<bool>? onChanged,

packages/flutter/lib/src/material/switch_list_tile.dart L349
(class SwitchListTile extends StatelessWidget)
-  final ValueChanged<bool?>? onChanged;
+  final ValueChanged<bool>? onChanged;

packages/flutter/lib/src/material/toggleable.dart L39
(abstract class RenderToggleable extends RenderConstrainedBox)
-    ValueChanged<bool?>? onChanged,
+    ValueChanged<bool>? onChanged,

packages/flutter/lib/src/material/toggleable.dart L343
(abstract class RenderToggleable extends RenderConstrainedBox)
-  ValueChanged<bool?>? get onChanged => _onChanged;
-  ValueChanged<bool?>? _onChanged;
-  set onChanged(ValueChanged<bool?>? value) {
+  ValueChanged<bool>? get onChanged => _onChanged;
+  ValueChanged<bool>? _onChanged;
+  set onChanged(ValueChanged<bool>? value) {
```

**Comment by goderbauer** [comment]
I believe RenderToggleable needs to keep its current signature to support the tri-state nature that it supports. However, the entities build on top of it (like Switch) should hide the tri-state if they don't support it.

**Comment by pennzht** [comment]
The tricky situation is:

* `bool` is a proper subtype of `bool?`
* Therefore, `ValueChanged<bool>` is a proper **supertype** of `ValueChanged<bool?>`
* So `Switch` cannot directly pass its `ValueChanged<bool>? onChanged` to `RenderToggleable`'s `ValueChanged<bool?>? onChanged`

```dart
typedef ValueChanged<T> = void Function(T value);
```

**Comment by darrenaustin** [comment]
> * So `Switch` cannot directly pass its `ValueChanged<bool>? onChanged` to `RenderToggleable`'s `ValueChanged<bool?>? onChanged`

That's why this PR wraps the `ValueChanged<bool>?` param passed to the `Switch` widget with the `_handleValueChanged(bool? value)` that it then passes to the lower level widgets (which eventually creates a `RenderToggleable` with this callback).

As Michael pointed out in his review we should probably take this a couple of steps further and change the type of the callback for everything below the `Switch` widget but above the `RenderToggleable`, as it still needs to support tri-state widgets.

**Comment by darrenaustin** [comment]
As both the Switch and Slider tests were more involved and required changes to widget framework code, I have removed them from this PR and will address them in separate PRs.

Reviewers, can you take a look at this reduced set of changes again?

**Review by goderbauer — APPROVED** [review]
LGTM

**Pull request merged**